File: ports/elixir/lib/unicode_security/form/normalization_bomb.ex
defmodule UnicodeSecurity.Form.NormalizationBomb do
  alias UnicodeSecurity.Ucd

  @max_nfkd_per_cp 8
  @nfd_ratio_pct 300
  @nfkd_ratio_pct 400

  def detect(input) do
    cond do
      (pos = first_blowup_cp(input)) != nil ->
        %{sub: "SingleCpBlowup", positions: [pos]}

      ratio_pct(input, &Ucd.to_nfkd/1) > @nfkd_ratio_pct ->
        %{sub: "NfkdHighExpansion", positions: []}

      ratio_pct(input, &Ucd.to_nfd/1) > @nfd_ratio_pct ->
        %{sub: "NfdHighExpansion", positions: []}

      true ->
        %{sub: nil, positions: []}
    end
  end

  defp first_blowup_cp(input),
    do:
      input
      |> Enum.with_index()
      |> Enum.find_value(fn {cp, i} ->
        if length(Ucd.to_nfkd([cp])) > @max_nfkd_per_cp, do: i, else: nil
      end)

  defp ratio_pct([], _fun), do: 0
  defp ratio_pct(input, fun), do: div(length(fun.(input)) * 100, length(input))
end

File: ports/lua/src/unicode_lua/security/form/normalization_bomb.lua
local ucd = require("unicode_lua.security.identity.ucd")

local M = {}

local MAX_NFKD_PER_CP = 8
local NFD_RATIO_PCT = 300
local NFKD_RATIO_PCT = 400

local function first_blowup_cp(input)
  for i = 1, #input do
    if #ucd.to_nfkd({ input[i] }) > MAX_NFKD_PER_CP then
      return i - 1
    end
  end
  return nil
end

local function nfd_ratio_pct(input)
  if #input == 0 then
    return 0
  end
  return math.floor(#ucd.to_nfd(input) * 100 / #input)
end

local function nfkd_ratio_pct(input)
  if #input == 0 then
    return 0
  end
  return math.floor(#ucd.to_nfkd(input) * 100 / #input)
end

function M.detect(input)
  local pos = first_blowup_cp(input)
  if pos ~= nil then
    return { sub = "SingleCpBlowup", positions = { pos } }
  end
  if nfkd_ratio_pct(input) > NFKD_RATIO_PCT then
    return { sub = "NfkdHighExpansion", positions = {} }
  end
  if nfd_ratio_pct(input) > NFD_RATIO_PCT then
    return { sub = "NfdHighExpansion", positions = {} }
  end
  return { sub = nil, positions = {} }
end

return M
